### ISSUE TITLE
Deleted

### ISSUE THREAD

**Issue by NicolasAlmerge** [pyparsing/plusminus#3]
Deleted

**Comment by ptmcg** [comment]
Thanks for trying out plusminus!

I think that was the behavior in an earlier release - the latest 0.3.0 release supports 6 levels of ()'s. When you get to 7, you'll get:

```
(((((((6)))))))
OverflowError: expression too deeply nested
```

The latest version should be runnable at http://ptmcg.pythonanywhere.com/plusminus - if there is a discrepancy between the 0.3.0 release and the behavior on this site, then I'll need to check into why that is.

**Comment by TheOneMusic** [comment]
Ok I see, how can I download the new release?

Is there a way to bypass the `RecursionError` when using only parenthesis and not functions? The [`founFn.py`](https://github.com/pyparsing/pyparsing/blob/master/examples/fourFn.py) example doesn't have this problem.

Thank you for making [`plusminus`](https://github.com/pyparsing/plusminus)!

**Comment by TheOneMusic** [comment]
Okay I see, how can I download the new release?

Is there a way to bypass the `RecursionError` when using only parenthesis and not functions? The [`founFn.py`](https://github.com/pyparsing/pyparsing/blob/master/examples/fourFn.py) example doesn't have this problem.

Thank you for making [`plusminus`](https://github.com/pyparsing/plusminus)!

**Comment by ptmcg** [comment]
How did you install it in the first place? plusminus is on PyPI, so `python -m pip install -U plusminus` should do it, I think

**Comment by TheOneMusic** [comment]
That's the way I installed it, without the `-U`

**Comment by ptmcg** [comment]
`-U` is for "update"

**Comment by TheOneMusic** [comment]
I uninstalled and reinstalled but nothing changed. I already had the latest version.

**Comment by TheOneMusic** [comment]
Also, is there a way to bypass the `RecursionError`/`OverflowError` when using only parenthesis and not functions? The [`founFn.py`](https://github.com/pyparsing/pyparsing/blob/master/examples/fourFn.py) example doesn't have this problem.

**Comment by ptmcg** [comment]
Ah, now I remember. Python has a default recursion limit of 1000, but arithmetic parsers created with pyparsing can exceed this number. The online parser demo includes this code:

    import sys
    sys.setrecursionlimit(3000)

Be judicious in updating this value.

`plusminus` uses a different mechanism than `fourFn.py` for building its parser. The `plusminus` parser also has an attribute (`maximum_expression_depth`) for changing this limitation to a larger number, but realize that larger values can end up creating deeply nested function calls, which can become a kind of denial-of-service of your application.

Between updating the Python recursion limit and increasing the maximum_expression_depth, I think you will be able handle more deeply-nested expressions. (It will still choke on expressions like "((((((((((((((((((((((((((((((((((((((((1))))))))))))))))))))))))))))))))))))))", I expect though.)

**Comment by ptmcg** [comment]
I can see I need to update my plusminus docs.

Out of curiosity, if you can tell me, what kind of application are you writing that uses this module?  It is still very early in its lifetime, and it helps me to know how it is being used when determining features and changes.

You might also find that BasicArithmeticParser has some "bells and whistles" that you don't necessarily care for or need - you can try using just the ArithmeticParser base class.

**Comment by TheOneMusic** [comment]
I need the application for a command for a discord bot. I have built my own parser, being inspired from the [`founFn.py`](https://github.com/pyparsing/pyparsing/blob/master/examples/fourFn.py) example, but I now wonder what I should use. The [`plusminus`](https://github.com/pyparsing/plusminus) package is great, but there are other advantages with the other parser, such as no errors when using lots of parenthesis (which can represent only mathematical expressions). For example, parsing `"((((((((((((((((((((((((((((((((((((((((6))))))))))))))))))))))))))))))))))))))))"` works and is fast.

Sorry for the delay, I was sleeping.

**Comment by TheOneMusic** [comment]
The `ArithmeticParser` class has a similar problem:

```python
>>> from plusminus import ArithmeticParser
>>> parser = ArithmeticParser()
>>> parser.evaluate("(((((6)))))") # 5 pairs of ()
6
>>> parser.evaluate("((((((6))))))") # 6 pairs of ()
RecursionError: maximum recursion depth exceeded
>>> parser.evaluate("(((((((6)))))))") # 7 pairs of ()
OverflowError: expression too deeply nested
```

**Comment by TheOneMusic** [comment]
Also, I copied the [`plusminus`](https://github.com/pyparsing/plusminus/) code and wrote it on another file to try some changes. Here is one change I could think of: allow functions to have optional arguments. For example, the [`log`](https://docs.python.org/3/library/math.html#math.log) function can be parsed using one or two arguments (the second argument is optional and represents the base, default being `10` in the example below, and not [`e`](https://docs.python.org/3/library/math.html#math.e)). Here is some changes:

```python
# plusminus_copy

class ArithmeticParser:

    ...

    class ArithmeticFunction(ArithNode):
        fn_map = None
        def evaluate(self):
            with _trimming_exception_traceback():
                fn_name, *fn_args = self.tokens
                if fn_name not in self.fn_map:
                    raise ValueError(
                        "{!r} is not a recognized function".format(fn_name)
                    )
                fn_spec = self.fn_map[fn_name]

                if isinstance(fn_spec.arity, tuple):
                    if not any(arit == len(fn_args) for arit in fn_spec.arity):
                        raise TypeError(
                    "{} takes {} {}, {} given".format(
                            fn_name,
                            " or ".join(str(ari) for ari in fn_spec.arity),
                            ("arg", "args")[fn_spec.arity[-1] != 1],
                            len(fn_args),
                        )
                    )

                elif fn_spec.arity not in (len(fn_args), ...):
                    raise TypeError(
                        "{} takes {} {}, {} given".format(
                            fn_name,
                            fn_spec.arity,
                            ("arg", "args")[fn_spec.arity != 1],
                            len(fn_args),
                        )
                    )
                return fn_spec.method(*[arg.evaluate() for arg in fn_args])

    ...

def log(x, y=10):
    if math.isclose(y, 2, abs_tol=1e-15):
        return math.log2(x)
    if math.isclose(y, 10, abs_tol=1e-15):
        return math.log10(x)
    return math.log(x, y)
```

```python
>>> from plusminus_copy import BasicArithmeticParser as BAP, log
>>> parser = BAP()
>>> parser.add_function("log", (1, 2), log)
>>> log(9)
0.954242509439325

>>> log(9, 10)
0.954242509439325

>>> log(9, 6)
1.226294385530917

>>> log(9, 7, 0)
TypeError: log takes 1 or 2 args, 3 given
```

**Comment by TheOneMusic** [comment]
Also, I copied the [`plusminus`](https://github.com/pyparsing/plusminus/) code and wrote it on another file to try some changes. Here is one change I could think of: allow functions to have optional arguments. For example, the [`log`](https://docs.python.org/3/library/math.html#math.log) function can be parsed using one or two arguments (the second argument is optional and represents the base, default being `10` in the example below, and not [`e`](https://docs.python.org/3/library/math.html#math.e)). Here are some changes I made:

```python
# plusminus_copy

...

class ArithmeticParser:

    ...

    class ArithmeticFunction(ArithNode):
        fn_map = None
        def evaluate(self):
            with _trimming_exception_traceback():
                fn_name, *fn_args = self.tokens
                if fn_name not in self.fn_map:
                    raise ValueError(
                        "{!r} is not a recognized function".format(fn_name)
                    )
                fn_spec = self.fn_map[fn_name]

                if isinstance(fn_spec.arity, tuple):
                    if not any(arit == len(fn_args) for arit in fn_spec.arity):
                        raise TypeError(
                    "{} takes {} {}, {} given".format(
                            fn_name,
                            " or ".join(str(ari) for ari in fn_spec.arity),
                            ("arg", "args")[fn_spec.arity[-1] != 1],
                            len(fn_args),
                        )
                    )

                elif fn_spec.arity not in (len(fn_args), ...):
                    raise TypeError(
                        "{} takes {} {}, {} given".format(
                            fn_name,
                            fn_spec.arity,
                            ("arg", "args")[fn_spec.arity != 1],
                            len(fn_args),
                        )
                    )
                return fn_spec.method(*[arg.evaluate() for arg in fn_args])

        ...

    ...

def log(x, y=10):
    if math.isclose(y, 2, abs_tol=1e-15):
        return math.log2(x)
    if math.isclose(y, 10, abs_tol=1e-15):
        return math.log10(x)
    return math.log(x, y)

...
```

```python
>>> from plusminus_copy import BasicArithmeticParser as BAP, log
>>> parser = BAP()
>>> parser.add_function("log", (1, 2), log)
>>> parser.evaluate("log(9)")
0.954242509439325
>>> parser.evaluate("log(9, 10)")
0.954242509439325
>>> parser.evaluate("log(9, 6)")
1.226294385530917
>>> parser.evaluate("log(9, 7, 0)")
TypeError: log takes 1 or 2 args, 3 given
```

**Comment by ptmcg** [comment]
What do you think about `plusminus` automatically raising the recursion limit when it gets imported?

**Comment by TheOneMusic** [comment]
@ptmcg What do you mean?

**Comment by ptmcg** [comment]
I think this issue gets resolved by adding `sys.setrecursionlimit(3000)` in your code. What do you think about `plusminus` doing that automatically when it gets imported? (it doesn't do that now, I'm thinking of adding it)

**Comment by TheOneMusic** [comment]
Oh okay, well why not, I think this might be good to avoid `RecursionError` when parsing an expression with depth of 6 (which is supposed to be the authorised). Can we **not** count single pairs of brackets in the expression's depth?

**Comment by TheOneMusic** [comment]
@ptmcg

**Comment by ptmcg** [comment]
Fixed in 0.4.0
